### PR TITLE
Port EditGCodeDialog from PrusaSlicer

### PR DESCRIPTION
**Pull request by Ocraftyone**
This is a port of the EditGCodeDialog from PrusaSlicer 2.7.x. There were a few changes made to make it a bit more functional. Also, it isn't quite fully complete, but it should be in a very usable state.

General Changes:
- Implement UndoValueUIManager and EditValueUI in Field
- Implement EditGCodeDialog and add buttons to the tabs
- Other minor changes to accommodate the new classes

Differences from PrusaSlicer's Implementation:
- backported to wxWidgets 3.1.5 (reverse commit 8770c4b7 after updating to 3.2.x)
- icons have been updated to use Orca colors
- improve the report that tells you if certain placeholders have not been defined properly for the dialog. It now shows all issues at once rather than having to fix then recompile to see the next issue.
- allow the use of the cmake option `ORCA_CHECK_GCODE_PLACEHOLDERS` to toggle the above report since our workflow rarely uses debug mode.
- if a custom gcode value is not set when checking gcode placeholders, a testing value is set. Custom gcode is not parsed if it is empty, and the only way to check if the placeholders are all defined is by running the placeholder operation on the custom gcode.
- some calls to `print.config()` in Gcode.cpp were changed to `m_config` to support the above testing values feature (only m_config is modified with the placeholders and `print.config()` would return an empty string)
- a macro has been added to quickly add a definition to SlicingStatesConfigDefs with less boiler plate (it could technically be used for any ConfigOptionDef, but that would hurt interoperability with PS. I tried to not use the macro for too many PS defined definitions.)
- the presets are now also categorized by the page they are on in their tab
<table>
<tr>
 <td>Prusa
 <td>Orca
<tr>
 <td>

 <td>

</table>

TODO:
- [x] Make sure all linux fixes have been applied
- [x] Finish adding "universal" gcode options
- [x] add search function to dataview (maybe?)
- [x] determine if any options are being left out of the preset categories by getting options from Tab rather than Presets. If so, consider adding outside of the groupings

### CODE REVIEW

**Comment by Ocraftyone** [comment]
sorry about the couple of extra PRs created. I had merged some of the changes from the wx upgrade ahead of time and it cause issues because of the reversion. Some cherry picking magic later, new branch!

**Comment by tsmith35** [comment]
@Ocraftyone and @SoftFever, I noticed your check builds were failing. I ran into the same issue here today and found that it's because the site hosting Boost (https://boostorg.jfrog.io/artifactory/main/release/1.78.0/source/boost_1_78_0.tar.gz) was returning "409 Conflict". Now it's prompting for the original registrant's email address to be added.

TL;DR, I got a copy of Boost from here: https://sourceforge.net/projects/boost/files/boost/1.78.0/boost_1_78_0.tar.gz

I put it in the OrcaSlicer\deps\DL_CACHE\Boost directory and after that, I could successfully compile. Just a PITA. lol

**Comment by SoftFever** [comment]
> @Ocraftyone and @SoftFever, I noticed your check builds were failing. I ran into the same issue here today and found that it's because the site hosting Boost (https://boostorg.jfrog.io/artifactory/main/release/1.78.0/source/boost_1_78_0.tar.gz) was returning "409 Conflict". Now it's prompting for the original registrant's email address to be added.
> 
> TL;DR, I got a copy of Boost from here: https://sourceforge.net/projects/boost/files/boost/1.78.0/boost_1_78_0.tar.gz
> 
> I put it in the OrcaSlicer\deps\DL_CACHE\Boost directory and after that, I could successfully compile. Just a PITA. lol

yeah, we noticed that too.
Lmao, let's see how long boost.org will recover it. If not, we will switch to github

**Comment by Ocraftyone** [comment]
Just curious, why did we not use GitHub in the first place? Seems to be a very reliable source and more well known.

**Comment by Ocraftyone** [comment]
Just curious, why did we not use GitHub in the first place? Seems to be a very reliable source and more well known.

**Comment by SoftFever** [comment]
> Just curious, why did we not use GitHub in the first place? Seems to be a very reliable source and more well known.

Because Boost doesn't use github and we just download from their official links.

**Comment by Ocraftyone** [comment]
I guess it doesn't matter because the link is back up! 🎉

**Comment by SoftFever** [comment]
> I guess it doesn't matter because the link is back up! 🎉

lmao, just 1 hour after I changed to github cache

**Comment by tsmith35** [comment]
> > I guess it doesn't matter because the link is back up!
> 
> lmao, just 1 hour after I changed to github cache

That is expected behavior, much like finding a lost item after buying a replacement. 😁

**Comment by Ocraftyone** [comment]
@SoftFever I believe we are good to go on the features. I have tested a bit in windows and I am about to do some testing in ubuntu. @Noisyfox , @discip , @igiannakas , and any other testers, if you don't mind giving it a quick overview and LMK if there are any improvements or bugs to address, that would be awesome!

**Comment by discip** [comment]
Good morning @Ocraftyone,
First and foremost: Thank you for this contribution!
If you wouldn't have done this I would probably never now there was a g-code-editor in Orca. 😅

The only thing I noticed using Windows:
The plus symbol in the middle jumps up and down when selecting different entries on the left side.

Will test Linux later.

**Comment by Ocraftyone** [comment]
I see what is happening. It is when the description text goes longer than a single line an it wraps. I will see if I can hardcode a position for the button to try and prevent that during resizing

> If you wouldn't have done this I would probably never now there was a g-code-editor in Orca. 😅

There isn't. Now there is 😁

**Comment by Ocraftyone** [comment]
@discip I got that change implemented. It now is positioned half way down the original height of the parameter list. This should be fine for now, but I did also identify another bug in the process. The window is resizable and the text field properly resizes, but the param list does not resize vertically. I am thinking that this is due to a wxWidgets change though because I am not able to identify any code that relates to that.

**Comment by Ocraftyone** [comment]
Latest commit removes the ability to resize the border. It solves a few issues that are laid out in the commit message

**Comment by SoftFever** [comment]
> @SoftFever I believe we are good to go on the features. I have tested a bit in windows and I am about to do some testing in ubuntu. @Noisyfox , @discip , @igiannakas , and any other testers, if you don't mind giving it a quick overview and LMK if there are any improvements or bugs to address, that would be awesome!

Thanks @Ocraftyone !
I should be able to start to review PRs again from tomorrow. 
Was fully occupied with until today😅

**Comment by Ocraftyone** [comment]
> > @SoftFever I believe we are good to go on the features. I have tested a bit in windows and I am about to do some testing in ubuntu. @Noisyfox , @discip , @igiannakas , and any other testers, if you don't mind giving it a quick overview and LMK if there are any improvements or bugs to address, that would be awesome!
> 
> Thanks @Ocraftyone ! I should be able to start to review PRs again from tomorrow. Was fully occupied with until today😅

No problem! I am trying to do some more tests in the mean time. I would ideally like someone to test on mac too, which is why I pinged @igiannakas

**Comment by discip** [comment]
@Ocraftyone
Finally I had time to test this on Linux (Manjaro) but unfortunately it doesn't work at all.
As soon as I click on the corresponding icon, it asks me to either wait or to force quit, but even if I choose "Wait" it crashes.

Also I'm not sure where to find the related logfile.

**Comment by discip** [comment]
@Ocraftyone
Finally I had time to test this on Linux (Manjaro) but unfortunately it doesn't work at all.
As soon as I click on the corresponding icon, it asks me to either wait or to force quit, but even if I choose "Wait" it crashes.

Also I'm not sure where to find the related logfile.


<details>
  <summary>
    <b>setup</b>
  </summary>
<br>

CPU | GPU | RAM | SSD
-|-|-|-
Ryzen 9 7950X3D | XFX 7900 XTX | G.Skill Flare X5 32 GB (EXPO I enabled) | Samsung 990 Pro 2 TB

</details>

**Comment by Ocraftyone** [comment]
@igiannakas would mind trying PrusaSlicer and see if it does something similar?

**Comment by igiannakas** [comment]
> @igiannakas would mind trying PrusaSlicer and see if it does something similar?

Yeap it does exactly the same... It's cosmetic so no real issue, but was the only thing I could see!

Another thing to consider is possibly filtering the list by the options available to the current print model - for example ramming doesn't apply to BBL printers etc. Again, that is polishing type stuff.

**Comment by Ocraftyone** [comment]
> Another thing to consider is possibly filtering the list by the options available to the current print model - for example ramming doesn't apply to BBL printers etc. Again, that is polishing type stuff.

That would be nice, but it would be a pretty manual process of determining all of the options that don't apply to BBL printers. Unless there is somewhere that specifies the options that are disabled or there are only a handful of options that would need to be disabled, I am hesitant to do that.

As for the dataview issue, do you have the same issue with the object list dataview? Maybe it has some logic to alleviate this. While building this, I saw that there were some options indicating that you can set an animation for expanding and collapsing items, but I did not mess with them. Maybe it has to be explicitly disabled for mac.

**Comment by Ocraftyone** [comment]
> Finally I had time to test this on Linux (Manjaro) but unfortunately it doesn't work at all.
> As soon as I click on the corresponding icon, it asks me to either wait or to force quit, but even if I choose "Wait" it crashes.

Just got a chance to test on ubuntu and I am getting a seg fault. I am having trouble building it in my test environment with github codespaces, so I still haven't been able to debug it yet.

File: src/slic3r/GUI/OG_CustomCtrl.cpp [hunk 1]
```diff
@@ -373,22 +373,31 @@ void OG_CustomCtrl::OnMotion(wxMouseEvent& event)
             break;
         }
 
-        for (size_t opt_idx = 0; opt_idx < line.rects_undo_icon.size(); opt_idx++)
+        size_t undo_icons_cnt = line.rects_undo_icon.size();
```
Comment (SoftFever) [inline src/slic3r/GUI/OG_CustomCtrl.cpp:376]: Curiously, what are these changes for?

File: src/slic3r/GUI/Field.hpp [hunk 2]
```diff
@@ -41,7 +41,120 @@ wxString double_to_string(double const value, const int max_precision = 4);
 wxString get_thumbnail_string(const Vec2d& value);
 wxString get_thumbnails_string(const std::vector<Vec2d>& values);
 
-class Field {
+class UndoValueUIManager
```
Comment (SoftFever) [inline src/slic3r/GUI/Field.hpp:44]: Just trying to understand here: if I'm not wrong, this micro refactoring is not directly related to the EditGCodeDialog, right?

File: src/slic3r/GUI/OptionsGroup.hpp [hunk 3]
```diff
@@ -155,6 +156,14 @@ class OptionsGroup {
 							if (m_fields.find(id) == m_fields.end()) return nullptr;
 							return m_fields.at(id).get();
     }
+
+    inline Line*	get_line(const t_config_option_key& id) {
+		for (Line& line : m_lines)
+			if (line.has_only_option(id))
```
Comment (SoftFever) [inline src/slic3r/GUI/OptionsGroup.hpp:162]: Shall we is_separator() too?
I can't recall the exact causes why I added that check. I remember there were crashes caused by separators

**Comment by Ocraftyone** [comment]
@discip I fixed that linux crash. It was one of the STUPIDEST mistake I have ever made, but its fixed now! 🤦

File: src/slic3r/GUI/OG_CustomCtrl.cpp (hunk 1 above)
Comment (Ocraftyone) [reply src/slic3r/GUI/OG_CustomCtrl.cpp:376]: I believe a lot of it is a refactor. Rather than having multiple loops, it is a single loop that handles all cases

File: src/slic3r/GUI/Field.hpp (hunk 2 above)
Comment (Ocraftyone) [reply src/slic3r/GUI/Field.hpp:44]: According to the commits from PS, it was a refactoring done in preparation for the dialog. It is still fully compatible. The EditUIManager IS directly related.

File: src/slic3r/GUI/OptionsGroup.hpp (hunk 3 above)
Comment (Ocraftyone) [reply src/slic3r/GUI/OptionsGroup.hpp:162]: ill look back into the commit it was added and get back to you

Comment (Ocraftyone) [reply src/slic3r/GUI/OptionsGroup.hpp:162]: The commit is from the implementation of internal bridge speed. I didn't get any crashes, but that doesn't mean much. I am kinda leaning towards just going back the the original BBL function...

**Comment by tsmith35** [comment]
> @discip I fixed that linux crash. It was one of the STUPIDEST mistake I have ever made, but its fixed now! 🤦

People who don't make mistakes never learn anything and are most likely doing nothing. You're doing fine. 👍

**Comment by Ocraftyone** [comment]
> > @discip I fixed that linux crash. It was one of the STUPIDEST mistake I have ever made, but its fixed now! 🤦
> 
> @Ocraftyone People who don't make mistakes never learn anything and are most likely doing nothing. You're doing fine. 👍

Ain't that the truth!

**Comment by discip** [comment]
@Ocraftyone
Sorry for the late answer. 😅

Indeed, it works now on Linux as well, but I think the color of the text is definitely too dim:

image

Just that.
Otherwise: Much appreciated!

File: src/slic3r/GUI/OptionsGroup.hpp (hunk 3 above)
Comment (SoftFever) [reply src/slic3r/GUI/OptionsGroup.hpp]: hehehe, that change was introduced by myself.
It was crashing when a separator line was added

**Review by SoftFever — APPROVED** [review]
Fantastic work!
Thank you very much

**Comment by discip** [comment]
@Ocraftyone
@SoftFever

Sorry to bother, but this issue unfortunately persists.

> the color of the text is definitely too dim:
> 
> image

**Comment by discip** [comment]
@Ocraftyone
@SoftFever

Sorry to bother, but unfortunately this issue persists.

> the color of the text is definitely too dim:
> 
> image

<details>
  <summary>
    <b>setup</b>
  </summary>

CPU | GPU | RAM | SSD
-|-|-|-
Ryzen 9 7950X3D | XFX 7900 XTX | G.Skill Flare X5 32 GB (EXPO I enabled) | Samsung 990 Pro 2 TB

</details>

**Comment by discip** [comment]
@Ocraftyone
@SoftFever

Sorry to bother, but unfortunately this issue persists.

> the color of the text is definitely too dim:
> 
> image

<details>
  <summary>
    <b>setup</b>
  </summary>
<br>

CPU | GPU | RAM | SSD
-|-|-|-
Ryzen 9 7950X3D | XFX 7900 XTX | G.Skill Flare X5 32 GB (EXPO I enabled) | Samsung 990 Pro 2 TB

</details>

**Comment by Ocraftyone** [comment]
@discip I'll take a look when I get a chance. I had meant to before merging but I will try to get a patch together.

**Comment by discip** [comment]
## Thank you! 👍🏻

I just noticed there are even more cases where the font is to dark:
image

image

Maybe there are even more. 🤷🏻‍♂️

**Comment by discip** [comment]
## Thank you! 👍🏻

I just noticed there are even more cases where the font is to dark:
image

image

Maybe there is even more. 🤷🏻‍♂️

**Comment by discip** [comment]
@Ocraftyone
Have you had time to look into that by any chance?

**Comment by tsmith35** [comment]
> @Ocraftyone Have you had time to look into that by any chance?

Is the text really that dark on a Mac? Is there a light mode or similar available?

**Comment by discip** [comment]
This is a screenshot. 😆
It shows what I see on my machine. 😊

In the linux version there is no option to choose from.
